mbed_flasher/flashers/FlasherJLink.py

```diff
@@ -0,0 +1,113 @@
+    supported_targets = None
+
+    def __init__(self, logger=None):
+        self.logger = logger if logger else logging.getLogger("mbed-flasher")
+
+    @staticmethod
+    def get_supported_targets():
+        """
+        :return: supported JLink types
+        """
+        if not FlasherJLink.supported_targets:
+            mbeds = mbed_lstools.create()
+            # @todo this is workaround until mbed-ls provide public api
+            db_items = list(mbeds.plat_db.items(device_type='jlink'))
+            FlasherJLink.supported_targets = sorted([i[1]["platform_name"] for i in db_items])
+
+        return FlasherJLink.supported_targets
+
+    @staticmethod
+    def get_available_devices():
+        """
+        :return: list of available devices
+        """
+        mbeds = mbed_lstools.create()
+        # device_type introduced in mbedls version 1.4.0
+        return mbeds.list_mbeds(filter_function=lambda m: m['device_type'] == 'jlink')
+
+    @staticmethod
+    def is_jlink_installed():
+        return spawn.find_executable(FlasherJLink.executable) is not None
+
+    def flash(self, source, target, method, no_reset):
+        """flash device
+        :param source: binary to be flashed
+        :param target: target to be flashed
+        :param method: method to use when flashing
+        :param no_reset: do not reset flashed board at all
+        :return: 0 when flashing success
+        """
+
+        cmd_script = tempfile.NamedTemporaryFile()
+        # JLinkExe flash has a check for binary content, erase
+        # before flash to disable it.
+        cmd_script.write('erase\n')
+        cmd_script.write('loadfile {}\n'.format(source))
+        if no_reset:
+            cmd_script.write('r\n')
+            cmd_script.write('g\n')
+        cmd_script.write('exit\n')
+        cmd_script.flush()
+
+        try:
+            args = [
+                FlasherJLink.executable,
+                "-if", "swd",
+                "-speed", "auto",
+                "-device", target["jlink_device_name"],
+                "-commanderscript", cmd_script.name
+            ]
+        except KeyError:
+            self.logger.exception("jlink_device_name missing from target")
+            return EXIT_CODE_FLASH_FAILED
+
+        try:
+            self.logger.info("Flashing {} with command {}".format(target["target_id"], args))
+            output = subprocess.check_output(args, stderr=subprocess.STDOUT)
+            self.logger.info("Flash of {} succeeded".format(target["target_id"]))
+            self.logger.debug(output)
+            return EXIT_CODE_SUCCESS
+        except subprocess.CalledProcessError as error:
+            self.logger.error("Flash of {} failed, with returncode {}"
+                              .format(target["target_id"], error.returncode))
+            self.logger.error(error.output)
+            return EXIT_CODE_FLASH_FAILED
+        finally:
+            cmd_script.close()
```

Comment:
Is there some reason this is not closed earlier?

Reply:
From NamedTemporaryFile docs:
 If delete is true (the default), the file is deleted as soon as it is closed.